{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# step1 : sra2fastq文件"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import os\n",
    "from concurrent.futures import ThreadPoolExecutor, as_completed\n",
    "\n",
    "## 批量下载SRA数据\n",
    "def prefetch_sra(sra_accession, work_dir):\n",
    "    try:\n",
    "        subprocess.run(['prefetch', '-O', work_dir, sra_accession], check=True)\n",
    "        print(f\"Prefetching {sra_accession} completed successfully.\")\n",
    "        return sra_accession\n",
    "    except subprocess.CalledProcessError as e:\n",
    "        print(f\"Prefetching {sra_accession} failed with error: {e}\")\n",
    "        return None\n",
    "\n",
    "## SRA转为fastq\n",
    "def fastq_dump(sra_accession, work_dir, output_dir):\n",
    "    sra_file = os.path.join(work_dir, sra_accession, f\"{sra_accession}.sra\")\n",
    "    if not os.path.exists(sra_file):\n",
    "        print(f\"SRA file {sra_file} does not exist. Skipping fastq-dump.\")\n",
    "        return\n",
    "    try:\n",
    "        subprocess.run(['fastq-dump', '--gzip', '--split-3', '-O', output_dir, sra_file], check=True)\n",
    "        print(f\"Fastq-dump {sra_accession} completed successfully.\")\n",
    "    except subprocess.CalledProcessError as e:\n",
    "        print(f\"Fastq-dump {sra_accession} failed with error: {e}\")\n",
    "\n",
    "def process_sra(sra_accession, work_dir, output_dir):\n",
    "    if prefetch_sra(sra_accession, work_dir):\n",
    "        fastq_dump(sra_accession, work_dir, output_dir)\n",
    "\n",
    "work_dir = \"/home/john/RNA-seq/RNASeqDemoData/airway\"\n",
    "output_dir = '/home/john/RNA-seq/RNASeqDemoData/fastq'\n",
    "max_workers = 10\n",
    "\n",
    "with open(\"/home/john/RNA-seq/RNASeqDemoData/airway/sra_id\", \"r\") as file:\n",
    "    sra_ids = [line.strip() for line in file if line.strip()]\n",
    "\n",
    "    with ThreadPoolExecutor(max_workers=max_workers) as executor:\n",
    "        futures = [executor.submit(process_sra, sra_id, work_dir, output_dir) for sra_id in sra_ids]\n",
    "        for future in as_completed(futures):\n",
    "            pass\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# step2 : 质量检测 (fastqc + multiqc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['HeSRD1_R1.fq.gz', 'HeSRA1_R2.fq.gz', 'HeSRT2_R1.fq.gz', 'HeSRT1_R2.fq.gz', 'HeSRD1_R2.fq.gz', 'HeSRA2_R2.fq.gz', 'HeSRD2_R1.fq.gz', 'HeSRT3_R2.fq.gz', 'HeSRD3_R2.fq.gz', 'HeSRT3_R1.fq.gz', 'HeSRA2_R1.fq.gz', 'HeSRD3_R1.fq.gz', 'HeSRT1_R1.fq.gz', 'HeSRD2_R2.fq.gz', 'HeSRA3_R2.fq.gz', 'HeSRA3_R1.fq.gz', 'HeSRA1_R1.fq.gz', 'HeSRT2_R2.fq.gz']\n",
      "FastQC has finished analyzing /home/john/ZouLab_Code/RNASeq/Rawdata/HeSR/HeSRD1_R1.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "xdg-open: file '/home/john/ZouLab_Code/RNASeq/Fastqc/HeSR_fastqc/SigAnal/HeSRD1_R1.fq.gz_fastqc.html' does not exist\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FastQC has finished analyzing /home/john/ZouLab_Code/RNASeq/Rawdata/HeSR/HeSRA1_R2.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "xdg-open: file '/home/john/ZouLab_Code/RNASeq/Fastqc/HeSR_fastqc/SigAnal/HeSRA1_R2.fq.gz_fastqc.html' does not exist\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FastQC has finished analyzing /home/john/ZouLab_Code/RNASeq/Rawdata/HeSR/HeSRT2_R1.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "xdg-open: file '/home/john/ZouLab_Code/RNASeq/Fastqc/HeSR_fastqc/SigAnal/HeSRT2_R1.fq.gz_fastqc.html' does not exist\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FastQC has finished analyzing /home/john/ZouLab_Code/RNASeq/Rawdata/HeSR/HeSRT1_R2.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "xdg-open: file '/home/john/ZouLab_Code/RNASeq/Fastqc/HeSR_fastqc/SigAnal/HeSRT1_R2.fq.gz_fastqc.html' does not exist\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FastQC has finished analyzing /home/john/ZouLab_Code/RNASeq/Rawdata/HeSR/HeSRD1_R2.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "xdg-open: file '/home/john/ZouLab_Code/RNASeq/Fastqc/HeSR_fastqc/SigAnal/HeSRD1_R2.fq.gz_fastqc.html' does not exist\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FastQC has finished analyzing /home/john/ZouLab_Code/RNASeq/Rawdata/HeSR/HeSRA2_R2.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "xdg-open: file '/home/john/ZouLab_Code/RNASeq/Fastqc/HeSR_fastqc/SigAnal/HeSRA2_R2.fq.gz_fastqc.html' does not exist\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FastQC has finished analyzing /home/john/ZouLab_Code/RNASeq/Rawdata/HeSR/HeSRD2_R1.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "xdg-open: file '/home/john/ZouLab_Code/RNASeq/Fastqc/HeSR_fastqc/SigAnal/HeSRD2_R1.fq.gz_fastqc.html' does not exist\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FastQC has finished analyzing /home/john/ZouLab_Code/RNASeq/Rawdata/HeSR/HeSRT3_R2.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "xdg-open: file '/home/john/ZouLab_Code/RNASeq/Fastqc/HeSR_fastqc/SigAnal/HeSRT3_R2.fq.gz_fastqc.html' does not exist\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FastQC has finished analyzing /home/john/ZouLab_Code/RNASeq/Rawdata/HeSR/HeSRD3_R2.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "xdg-open: file '/home/john/ZouLab_Code/RNASeq/Fastqc/HeSR_fastqc/SigAnal/HeSRD3_R2.fq.gz_fastqc.html' does not exist\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FastQC has finished analyzing /home/john/ZouLab_Code/RNASeq/Rawdata/HeSR/HeSRT3_R1.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "xdg-open: file '/home/john/ZouLab_Code/RNASeq/Fastqc/HeSR_fastqc/SigAnal/HeSRT3_R1.fq.gz_fastqc.html' does not exist\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FastQC has finished analyzing /home/john/ZouLab_Code/RNASeq/Rawdata/HeSR/HeSRA2_R1.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "xdg-open: file '/home/john/ZouLab_Code/RNASeq/Fastqc/HeSR_fastqc/SigAnal/HeSRA2_R1.fq.gz_fastqc.html' does not exist\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FastQC has finished analyzing /home/john/ZouLab_Code/RNASeq/Rawdata/HeSR/HeSRD3_R1.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "xdg-open: file '/home/john/ZouLab_Code/RNASeq/Fastqc/HeSR_fastqc/SigAnal/HeSRD3_R1.fq.gz_fastqc.html' does not exist\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FastQC has finished analyzing /home/john/ZouLab_Code/RNASeq/Rawdata/HeSR/HeSRT1_R1.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "xdg-open: file '/home/john/ZouLab_Code/RNASeq/Fastqc/HeSR_fastqc/SigAnal/HeSRT1_R1.fq.gz_fastqc.html' does not exist\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FastQC has finished analyzing /home/john/ZouLab_Code/RNASeq/Rawdata/HeSR/HeSRD2_R2.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "xdg-open: file '/home/john/ZouLab_Code/RNASeq/Fastqc/HeSR_fastqc/SigAnal/HeSRD2_R2.fq.gz_fastqc.html' does not exist\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FastQC has finished analyzing /home/john/ZouLab_Code/RNASeq/Rawdata/HeSR/HeSRA3_R2.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "xdg-open: file '/home/john/ZouLab_Code/RNASeq/Fastqc/HeSR_fastqc/SigAnal/HeSRA3_R2.fq.gz_fastqc.html' does not exist\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FastQC has finished analyzing /home/john/ZouLab_Code/RNASeq/Rawdata/HeSR/HeSRA3_R1.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "xdg-open: file '/home/john/ZouLab_Code/RNASeq/Fastqc/HeSR_fastqc/SigAnal/HeSRA3_R1.fq.gz_fastqc.html' does not exist\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FastQC has finished analyzing /home/john/ZouLab_Code/RNASeq/Rawdata/HeSR/HeSRA1_R1.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "xdg-open: file '/home/john/ZouLab_Code/RNASeq/Fastqc/HeSR_fastqc/SigAnal/HeSRA1_R1.fq.gz_fastqc.html' does not exist\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FastQC has finished analyzing /home/john/ZouLab_Code/RNASeq/Rawdata/HeSR/HeSRT2_R2.fq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "xdg-open: file '/home/john/ZouLab_Code/RNASeq/Fastqc/HeSR_fastqc/SigAnal/HeSRT2_R2.fq.gz_fastqc.html' does not exist\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MultiQC report generated successfully.\n"
     ]
    }
   ],
   "source": [
    "\n",
    "import subprocess\n",
    "import os\n",
    "import webbrowser\n",
    "\n",
    "#每个fq单独质控\n",
    "def run_fastqc(fastq_file, output_dir):\n",
    "    \n",
    "    if not os.path.exists(output_dir):\n",
    "        os.makedirs(output_dir)\n",
    "        \n",
    "    fastqc_command = ['fastqc', fastq_file, '-o', output_dir]\n",
    "    ## FastQC\n",
    "    try:\n",
    "        subprocess.run(fastqc_command, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
    "        print(f\"FastQC has finished analyzing {fastq_file}\")\n",
    "        html_report_path = os.path.join(output_dir, os.path.basename(fastq_file) + '_fastqc.html')\n",
    "        webbrowser.open('file://' + html_report_path )\n",
    "        \n",
    "    except subprocess.CalledProcessError as e:\n",
    "        print(f\"An error occurred while running FastQC on: {fastq_file}: {e}\")\n",
    "\n",
    "#总质控\n",
    "def generate_multiqc_report(fastqczip_path, output_dir):\n",
    "    multiqc_command = ['multiqc', fastqczip_path, '-o', output_dir]\n",
    "    \n",
    "    try:\n",
    "        subprocess.run(multiqc_command, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
    "        print(\"MultiQC report generated successfully.\")\n",
    "    except subprocess.CalledProcessError as e:\n",
    "        print(f\"An error occurred while generated the MultiQC report:{e}\")\n",
    "\n",
    "fastq_path = \"/home/john/ZouLab_Code/RNASeq/Rawdata/HeSR/\"\n",
    "fastq_files = [ff for ff in os.listdir(fastq_path) if ff.endswith('fq.gz')]\n",
    "print(fastq_files)\n",
    "fastqcAnal_path = \"/home/john/ZouLab_Code/RNASeq/Fastqc/HeSR_fastqc/SigAnal\"\n",
    "multiqcAnal_path = \"/home/john/ZouLab_Code/RNASeq/Fastqc/HeSR_fastqc/MulAnal\"\n",
    "# print(fastqc_zips)\n",
    "for fq in fastq_files:\n",
    "    fastq_file = fastq_path + fq\n",
    "    run_fastqc(fastq_file, fastqcAnal_path)\n",
    "\n",
    "generate_multiqc_report(fastqcAnal_path, multiqcAnal_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 过滤低质量数据 (trim_galore)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "igzip command line interface 2.31.0\n",
      "\n",
      "Multicore support not enabled. Proceeding with single-core trimming.\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 4.9\n",
      "single-core operation.\n",
      "igzip detected. Using igzip for decompressing\n",
      "\n",
      "Output will be written into the directory: /home/john/ZouLab_Code/RNASeq/TrimGaloreData/HeSR_val/\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> /home/john/ZouLab_Code/RNASeq/Rawdata/HeSR/HeSRD3_R1.fq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "Illumina\t121814\tAGATCGGAAGAGC\t1000000\t12.18\n",
      "smallRNA\t0\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Nextera\t0\tCTGTCTCTTATA\t1000000\t0.00\n",
      "Using Illumina adapter for trimming (count: 121814). Second best hit was smallRNA (count: 0)\n",
      "\n",
      "Writing report to '/home/john/ZouLab_Code/RNASeq/TrimGaloreData/HeSR_val/HeSRD3_R1.fq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/john/ZouLab_Code/RNASeq/Rawdata/HeSR/HeSRD3_R1.fq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.9\n",
      "Number of cores used for trimming: 1\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 3 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.9). Setting -j 1\n",
      "Writing final adapter and quality trimmed output to HeSRD3_R1_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /home/john/ZouLab_Code/RNASeq/Rawdata/HeSR/HeSRD3_R1.fq.gz <<< \n",
      "10000000 sequences processed\n",
      "This is cutadapt 4.9 with Python 3.8.19\n",
      "Command line parameters: -j 1 -e 0.1 -q 20 -O 3 -a AGATCGGAAGAGC /home/john/ZouLab_Code/RNASeq/Rawdata/HeSR/HeSRD3_R1.fq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Finished in 170.306 s (11.053 µs/read; 5.43 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              15,408,171\n",
      "Reads with adapters:                14,866,687 (96.5%)\n",
      "Reads written (passing filters):    15,408,171 (100.0%)\n",
      "\n",
      "Total basepairs processed: 2,311,225,650 bp\n",
      "Quality-trimmed:               2,660,106 bp (0.1%)\n",
      "Total written (filtered):    152,096,692 bp (6.6%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 14866687 times\n",
      "\n",
      "Minimum overlap: 3\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 1.3%\n",
      "  C: 3.8%\n",
      "  G: 5.7%\n",
      "  T: 1.5%\n",
      "  none/other: 87.5%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t963\t240752.7\t0\t963\n",
      "4\t627\t60188.2\t0\t627\n",
      "5\t399\t15047.0\t0\t399\n",
      "6\t237\t3761.8\t0\t237\n",
      "7\t202\t940.4\t0\t202\n",
      "8\t212\t235.1\t0\t212\n",
      "9\t197\t58.8\t0\t196 1\n",
      "10\t269\t14.7\t1\t244 25\n",
      "11\t227\t3.7\t1\t205 22\n",
      "12\t257\t0.9\t1\t240 17\n",
      "13\t275\t0.2\t1\t254 21\n",
      "14\t283\t0.2\t1\t266 17\n",
      "15\t461\t0.2\t1\t435 26\n",
      "16\t496\t0.2\t1\t459 37\n",
      "17\t298\t0.2\t1\t266 32\n",
      "18\t299\t0.2\t1\t283 16\n",
      "19\t442\t0.2\t1\t415 27\n",
      "20\t459\t0.2\t1\t429 30\n",
      "21\t613\t0.2\t1\t565 48\n",
      "22\t381\t0.2\t1\t357 24\n",
      "23\t358\t0.2\t1\t328 30\n",
      "24\t381\t0.2\t1\t335 46\n",
      "25\t404\t0.2\t1\t326 78\n",
      "26\t467\t0.2\t1\t399 68\n",
      "27\t529\t0.2\t1\t440 89\n",
      "28\t511\t0.2\t1\t434 77\n",
      "29\t523\t0.2\t1\t468 55\n",
      "30\t537\t0.2\t1\t460 77\n",
      "31\t605\t0.2\t1\t503 102\n",
      "32\t743\t0.2\t1\t652 91\n",
      "33\t770\t0.2\t1\t677 93\n",
      "34\t721\t0.2\t1\t571 150\n",
      "35\t692\t0.2\t1\t543 149\n",
      "36\t943\t0.2\t1\t613 330\n",
      "37\t1077\t0.2\t1\t795 282\n",
      "38\t1048\t0.2\t1\t764 284\n",
      "39\t1375\t0.2\t1\t899 476\n",
      "40\t1637\t0.2\t1\t1124 513\n",
      "41\t2139\t0.2\t1\t1183 956\n",
      "42\t2610\t0.2\t1\t1913 697\n",
      "43\t1410\t0.2\t1\t837 573\n",
      "44\t1374\t0.2\t1\t961 413\n",
      "45\t1806\t0.2\t1\t1070 736\n",
      "46\t2651\t0.2\t1\t1767 884\n",
      "47\t3321\t0.2\t1\t2507 814\n",
      "48\t5138\t0.2\t1\t4526 612\n",
      "49\t4466\t0.2\t1\t3759 707\n",
      "50\t13355\t0.2\t1\t12736 619\n",
      "51\t12334\t0.2\t1\t11961 373\n",
      "52\t3532\t0.2\t1\t3265 267\n",
      "53\t9607\t0.2\t1\t9394 213\n",
      "54\t11554\t0.2\t1\t11343 211\n",
      "55\t4959\t0.2\t1\t4828 131\n",
      "56\t2008\t0.2\t1\t1944 64\n",
      "57\t1685\t0.2\t1\t1633 52\n",
      "58\t797\t0.2\t1\t757 40\n",
      "59\t963\t0.2\t1\t926 37\n",
      "60\t647\t0.2\t1\t603 44\n",
      "61\t359\t0.2\t1\t333 26\n",
      "62\t576\t0.2\t1\t542 34\n",
      "63\t1442\t0.2\t1\t1414 28\n",
      "64\t550\t0.2\t1\t520 30\n",
      "65\t688\t0.2\t1\t667 21\n",
      "66\t930\t0.2\t1\t897 33\n",
      "67\t1138\t0.2\t1\t1100 38\n",
      "68\t1617\t0.2\t1\t1566 51\n",
      "69\t1712\t0.2\t1\t1640 72\n",
      "70\t2608\t0.2\t1\t2535 73\n",
      "71\t3134\t0.2\t1\t3075 59\n",
      "72\t4410\t0.2\t1\t4321 89\n",
      "73\t5327\t0.2\t1\t5222 105\n",
      "74\t6869\t0.2\t1\t6706 163\n",
      "75\t6169\t0.2\t1\t6044 125\n",
      "76\t6346\t0.2\t1\t6167 179\n",
      "77\t7024\t0.2\t1\t6881 143\n",
      "78\t11382\t0.2\t1\t11136 246\n",
      "79\t7311\t0.2\t1\t7167 144\n",
      "80\t7435\t0.2\t1\t7274 161\n",
      "81\t7203\t0.2\t1\t7052 151\n",
      "82\t8105\t0.2\t1\t7925 180\n",
      "83\t7768\t0.2\t1\t7583 185\n",
      "84\t9919\t0.2\t1\t9696 223\n",
      "85\t9895\t0.2\t1\t9658 237\n",
      "86\t10798\t0.2\t1\t10558 240\n",
      "87\t14954\t0.2\t1\t14619 335\n",
      "88\t11740\t0.2\t1\t11535 205\n",
      "89\t12637\t0.2\t1\t12414 223\n",
      "90\t12011\t0.2\t1\t11765 246\n",
      "91\t12352\t0.2\t1\t12117 235\n",
      "92\t12743\t0.2\t1\t12480 263\n",
      "93\t11695\t0.2\t1\t11384 311\n",
      "94\t13657\t0.2\t1\t13401 256\n",
      "95\t13110\t0.2\t1\t12564 546\n",
      "96\t16221\t0.2\t1\t15880 341\n",
      "97\t17371\t0.2\t1\t16696 675\n",
      "98\t16575\t0.2\t1\t16245 330\n",
      "99\t21274\t0.2\t1\t20777 497\n",
      "100\t20689\t0.2\t1\t20283 406\n",
      "101\t21085\t0.2\t1\t20554 531\n",
      "102\t21367\t0.2\t1\t20911 456\n",
      "103\t20280\t0.2\t1\t19882 398\n",
      "104\t20817\t0.2\t1\t20377 440\n",
      "105\t22148\t0.2\t1\t21594 554\n",
      "106\t27742\t0.2\t1\t26996 746\n",
      "107\t28561\t0.2\t1\t27958 603\n",
      "108\t28897\t0.2\t1\t28191 706\n",
      "109\t33266\t0.2\t1\t32531 735\n",
      "110\t46088\t0.2\t1\t44782 1306\n",
      "111\t36281\t0.2\t1\t35319 962\n",
      "112\t83809\t0.2\t1\t81864 1945\n",
      "113\t39115\t0.2\t1\t37789 1326\n",
      "114\t89480\t0.2\t1\t85910 3570\n",
      "115\t39188\t0.2\t1\t38083 1105\n",
      "116\t48860\t0.2\t1\t47034 1826\n",
      "117\t75766\t0.2\t1\t73941 1825\n",
      "118\t42688\t0.2\t1\t41464 1224\n",
      "119\t55027\t0.2\t1\t53910 1117\n",
      "120\t45845\t0.2\t1\t44638 1207\n",
      "121\t47666\t0.2\t1\t46537 1129\n",
      "122\t103641\t0.2\t1\t100836 2805\n",
      "123\t35536\t0.2\t1\t34191 1345\n",
      "124\t32686\t0.2\t1\t31400 1286\n",
      "125\t28667\t0.2\t1\t27499 1168\n",
      "126\t20184\t0.2\t1\t19493 691\n",
      "127\t12450\t0.2\t1\t11921 529\n",
      "128\t14819\t0.2\t1\t14205 614\n",
      "129\t9707\t0.2\t1\t9349 358\n",
      "130\t4048\t0.2\t1\t3668 380\n",
      "131\t2202\t0.2\t1\t1865 337\n",
      "132\t1623\t0.2\t1\t1304 319\n",
      "133\t2638\t0.2\t1\t1960 678\n",
      "134\t11466\t0.2\t1\t5497 5969\n",
      "135\t37442\t0.2\t1\t36622 820\n",
      "136\t9990\t0.2\t1\t9465 525\n",
      "137\t6244\t0.2\t1\t5678 566\n",
      "138\t15092\t0.2\t1\t14220 872\n",
      "139\t6445\t0.2\t1\t5791 654\n",
      "140\t7917\t0.2\t1\t7161 756\n",
      "141\t29807\t0.2\t1\t28580 1227\n",
      "142\t5950\t0.2\t1\t5090 860\n",
      "143\t7110\t0.2\t1\t5923 1187\n",
      "144\t8379\t0.2\t1\t7014 1365\n",
      "145\t7996\t0.2\t1\t6130 1866\n",
      "146\t16826\t0.2\t1\t14632 2194\n",
      "147\t22975\t0.2\t1\t18265 4710\n",
      "148\t21485\t0.2\t1\t15307 6178\n",
      "149\t129277\t0.2\t1\t6857 122420\n",
      "150\t12935091\t0.2\t1\t235784 12699307\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/john/ZouLab_Code/RNASeq/Rawdata/HeSR/HeSRD3_R1.fq.gz\n",
      "=============================================\n",
      "15408171 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Writing report to '/home/john/ZouLab_Code/RNASeq/TrimGaloreData/HeSR_val/HeSRD3_R2.fq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/john/ZouLab_Code/RNASeq/Rawdata/HeSR/HeSRD3_R2.fq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.9\n",
      "Number of cores used for trimming: 1\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 3 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.9). Setting -j -j 1\n",
      "Writing final adapter and quality trimmed output to HeSRD3_R2_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /home/john/ZouLab_Code/RNASeq/Rawdata/HeSR/HeSRD3_R2.fq.gz <<< \n",
      "10000000 sequences processed\n",
      "This is cutadapt 4.9 with Python 3.8.19\n",
      "Command line parameters: -j 1 -e 0.1 -q 20 -O 3 -a AGATCGGAAGAGC /home/john/ZouLab_Code/RNASeq/Rawdata/HeSR/HeSRD3_R2.fq.gz\n",
      "Processing single-end reads on 1 core ...\n",
      "Finished in 73.612 s (4.777 µs/read; 12.56 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              15,408,171\n",
      "Reads with adapters:                 2,099,567 (13.6%)\n",
      "Reads written (passing filters):    15,408,171 (100.0%)\n",
      "\n",
      "Total basepairs processed: 2,311,225,650 bp\n",
      "Quality-trimmed:               8,756,403 bp (0.4%)\n",
      "Total written (filtered):  2,060,994,211 bp (89.2%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 2099567 times\n",
      "\n",
      "Minimum overlap: 3\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 7.6%\n",
      "  C: 29.2%\n",
      "  G: 43.0%\n",
      "  T: 7.6%\n",
      "  none/other: 12.7%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t1444\t240752.7\t0\t1444\n",
      "4\t391\t60188.2\t0\t391\n",
      "5\t244\t15047.0\t0\t244\n",
      "6\t241\t3761.8\t0\t241\n",
      "7\t215\t940.4\t0\t215\n",
      "8\t212\t235.1\t0\t212\n",
      "9\t232\t58.8\t0\t228 4\n",
      "10\t279\t14.7\t1\t257 22\n",
      "11\t274\t3.7\t1\t241 33\n",
      "12\t349\t0.9\t1\t310 39\n",
      "13\t258\t0.2\t1\t248 10\n",
      "14\t339\t0.2\t1\t311 28\n",
      "15\t522\t0.2\t1\t494 28\n",
      "16\t504\t0.2\t1\t461 43\n",
      "17\t376\t0.2\t1\t319 57\n",
      "18\t428\t0.2\t1\t380 48\n",
      "19\t1217\t0.2\t1\t1087 130\n",
      "20\t492\t0.2\t1\t449 43\n",
      "21\t267\t0.2\t1\t253 14\n",
      "22\t354\t0.2\t1\t325 29\n",
      "23\t416\t0.2\t1\t375 41\n",
      "24\t1003\t0.2\t1\t881 122\n",
      "25\t391\t0.2\t1\t349 42\n",
      "26\t657\t0.2\t1\t610 47\n",
      "27\t2267\t0.2\t1\t2138 129\n",
      "28\t672\t0.2\t1\t606 66\n",
      "29\t1325\t0.2\t1\t1259 66\n",
      "30\t992\t0.2\t1\t936 56\n",
      "31\t511\t0.2\t1\t484 27\n",
      "32\t907\t0.2\t1\t859 48\n",
      "33\t2775\t0.2\t1\t2710 65\n",
      "34\t660\t0.2\t1\t629 31\n",
      "35\t665\t0.2\t1\t644 21\n",
      "36\t2210\t0.2\t1\t2169 41\n",
      "37\t513\t0.2\t1\t500 13\n",
      "38\t1130\t0.2\t1\t1096 34\n",
      "39\t1149\t0.2\t1\t1109 40\n",
      "40\t1852\t0.2\t1\t1794 58\n",
      "41\t320\t0.2\t1\t300 20\n",
      "42\t1101\t0.2\t1\t1064 37\n",
      "43\t1699\t0.2\t1\t1665 34\n",
      "44\t833\t0.2\t1\t807 26\n",
      "45\t3113\t0.2\t1\t3031 82\n",
      "46\t4701\t0.2\t1\t4644 57\n",
      "47\t1044\t0.2\t1\t1014 30\n",
      "48\t1311\t0.2\t1\t1287 24\n",
      "49\t757\t0.2\t1\t722 35\n",
      "50\t3024\t0.2\t1\t2982 42\n",
      "51\t894\t0.2\t1\t852 42\n",
      "52\t2401\t0.2\t1\t2364 37\n",
      "53\t698\t0.2\t1\t679 19\n",
      "54\t1319\t0.2\t1\t1289 30\n",
      "55\t2912\t0.2\t1\t2845 67\n",
      "56\t1299\t0.2\t1\t1257 42\n",
      "57\t1113\t0.2\t1\t1052 61\n",
      "58\t4800\t0.2\t1\t4711 89\n",
      "59\t953\t0.2\t1\t902 51\n",
      "60\t3850\t0.2\t1\t3751 99\n",
      "61\t4433\t0.2\t1\t4355 78\n",
      "62\t5152\t0.2\t1\t5070 82\n",
      "63\t3920\t0.2\t1\t3841 79\n",
      "64\t963\t0.2\t1\t904 59\n",
      "65\t1563\t0.2\t1\t1513 50\n",
      "66\t2439\t0.2\t1\t2394 45\n",
      "67\t3460\t0.2\t1\t3387 73\n",
      "68\t5074\t0.2\t1\t4974 100\n",
      "69\t4764\t0.2\t1\t4665 99\n",
      "70\t5773\t0.2\t1\t5635 138\n",
      "71\t5008\t0.2\t1\t4907 101\n",
      "72\t5549\t0.2\t1\t5408 141\n",
      "73\t5995\t0.2\t1\t5872 123\n",
      "74\t7362\t0.2\t1\t7206 156\n",
      "75\t6343\t0.2\t1\t6192 151\n",
      "76\t6413\t0.2\t1\t6258 155\n",
      "77\t7077\t0.2\t1\t6932 145\n",
      "78\t11646\t0.2\t1\t11400 246\n",
      "79\t7343\t0.2\t1\t7183 160\n",
      "80\t7515\t0.2\t1\t7348 167\n",
      "81\t7236\t0.2\t1\t7096 140\n",
      "82\t8144\t0.2\t1\t7957 187\n",
      "83\t7747\t0.2\t1\t7582 165\n",
      "84\t9948\t0.2\t1\t9763 185\n",
      "85\t9933\t0.2\t1\t9722 211\n",
      "86\t10755\t0.2\t1\t10543 212\n",
      "87\t15240\t0.2\t1\t14902 338\n",
      "88\t11717\t0.2\t1\t11499 218\n",
      "89\t12711\t0.2\t1\t12473 238\n",
      "90\t12091\t0.2\t1\t11852 239\n",
      "91\t12528\t0.2\t1\t12301 227\n",
      "92\t12780\t0.2\t1\t12570 210\n",
      "93\t11619\t0.2\t1\t11413 206\n",
      "94\t13652\t0.2\t1\t13403 249\n",
      "95\t13044\t0.2\t1\t12796 248\n",
      "96\t16060\t0.2\t1\t15785 275\n",
      "97\t17838\t0.2\t1\t17523 315\n",
      "98\t16571\t0.2\t1\t16298 273\n",
      "99\t21447\t0.2\t1\t21135 312\n",
      "100\t20828\t0.2\t1\t20529 299\n",
      "101\t21467\t0.2\t1\t21102 365\n",
      "102\t21490\t0.2\t1\t21139 351\n",
      "103\t20517\t0.2\t1\t20193 324\n",
      "104\t21072\t0.2\t1\t20679 393\n",
      "105\t22857\t0.2\t1\t22392 465\n",
      "106\t27713\t0.2\t1\t27242 471\n",
      "107\t28862\t0.2\t1\t28422 440\n",
      "108\t28903\t0.2\t1\t28449 454\n",
      "109\t33619\t0.2\t1\t33075 544\n",
      "110\t46565\t0.2\t1\t45916 649\n",
      "111\t35751\t0.2\t1\t35161 590\n",
      "112\t84319\t0.2\t1\t83248 1071\n",
      "113\t38936\t0.2\t1\t38392 544\n",
      "114\t98165\t0.2\t1\t96732 1433\n",
      "115\t39755\t0.2\t1\t39135 620\n",
      "116\t52538\t0.2\t1\t51624 914\n",
      "117\t77063\t0.2\t1\t75990 1073\n",
      "118\t38501\t0.2\t1\t37897 604\n",
      "119\t54938\t0.2\t1\t53472 1466\n",
      "120\t54403\t0.2\t1\t53112 1291\n",
      "121\t45275\t0.2\t1\t43297 1978\n",
      "122\t105611\t0.2\t1\t103269 2342\n",
      "123\t39255\t0.2\t1\t38492 763\n",
      "124\t42669\t0.2\t1\t41562 1107\n",
      "125\t29951\t0.2\t1\t29483 468\n",
      "126\t21431\t0.2\t1\t20978 453\n",
      "127\t13114\t0.2\t1\t12889 225\n",
      "128\t15774\t0.2\t1\t15440 334\n",
      "129\t9724\t0.2\t1\t9573 151\n",
      "130\t4302\t0.2\t1\t4192 110\n",
      "131\t1882\t0.2\t1\t1849 33\n",
      "132\t1564\t0.2\t1\t1529 35\n",
      "133\t2125\t0.2\t1\t2077 48\n",
      "134\t6250\t0.2\t1\t6007 243\n",
      "135\t46573\t0.2\t1\t45680 893\n",
      "136\t13115\t0.2\t1\t12930 185\n",
      "137\t7127\t0.2\t1\t7006 121\n",
      "138\t20535\t0.2\t1\t20264 271\n",
      "139\t6027\t0.2\t1\t5939 88\n",
      "140\t7739\t0.2\t1\t7622 117\n",
      "141\t29287\t0.2\t1\t28952 335\n",
      "142\t5407\t0.2\t1\t5317 90\n",
      "143\t6146\t0.2\t1\t6062 84\n",
      "144\t7179\t0.2\t1\t7072 107\n",
      "145\t6296\t0.2\t1\t6164 132\n",
      "146\t14889\t0.2\t1\t14701 188\n",
      "147\t18692\t0.2\t1\t18365 327\n",
      "148\t15660\t0.2\t1\t15374 286\n",
      "149\t7886\t0.2\t1\t6825 1061\n",
      "150\t264092\t0.2\t1\t237179 26913\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/john/ZouLab_Code/RNASeq/Rawdata/HeSR/HeSRD3_R2.fq.gz\n",
      "=============================================\n",
      "15408171 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files HeSRD3_R1_trimmed.fq.gz and HeSRD3_R2_trimmed.fq.gz\n",
      "file_1: HeSRD3_R1_trimmed.fq.gz, file_2: HeSRD3_R2_trimmed.fq.gz\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: HeSRD3_R1_trimmed.fq.gz and HeSRD3_R2_trimmed.fq.gz <<<<<\n",
      "Writing validated paired-end Read 1 reads to HeSRD3_R1_val_1.fq.gz\n",
      "Writing validated paired-end Read 2 reads to HeSRD3_R2_val_2.fq.gz\n",
      "\n",
      "Total number of sequences analysed: 15408171\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 13324898 (86.48%)\n",
      "\n",
      "Deleting both intermediate output files HeSRD3_R1_trimmed.fq.gz and HeSRD3_R2_trimmed.fq.gz\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import subprocess\n",
    "\n",
    "# 定义工作目录和输出目录\n",
    "work_dir = '/home/john/ZouLab_Code/RNASeq/Rawdata/HeSR'\n",
    "output_dir = '/home/john/ZouLab_Code/RNASeq/TrimGaloreData/HeSR_val/'\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "# 查找所有符合条件的双端测序文件\n",
    "fq1_files = [file for file in os.listdir(work_dir) if 'HeSRD3_R1.fq.gz' in file]\n",
    "fq2_files = [file for file in os.listdir(work_dir) if 'HeSRD3_R2.fq.gz' in file]\n",
    "assert len(fq1_files) == len(fq2_files), \"Number of R1 and R2 files do not match.\"\n",
    "\n",
    "# 将文件对写入文件列表\n",
    "file_pairs = '\\n'.join(f'{fq1}\\t{fq2}' for fq1, fq2 in zip(fq1_files, fq2_files))\n",
    "with open('file_all', 'w') as f:\n",
    "    f.write(file_pairs)\n",
    "\n",
    "# 读取文件列表并处理每个文件对\n",
    "# with open('file_all', 'r') as f:\n",
    "#     for line in f:\n",
    "#         fq1, fq2 = line.strip().split()\n",
    "#         fq1_path = os.path.join(work_dir, fq1)\n",
    "#         fq2_path = os.path.join(work_dir, fq2)\n",
    "#         fq1_val_path = os.path.join(output_dir, fq1.replace('.fq.gz', '_cut.fq.gz'))\n",
    "#         fq2_val_path = os.path.join(output_dir, fq2.replace('.fq.gz', '_cut.fq.gz'))\n",
    "        \n",
    "#         print(f\"Processing: {fq1_path} and {fq2_path}\")\n",
    "        \n",
    "#         # 构建 cutadapt 命令\n",
    "#         cutadapt_command = [\n",
    "#             'cutadapt',\n",
    "#             '-a', 'AATG',  # Read1的3'端适配器\n",
    "#             '-A', 'CAAGCAGAAGACGGCATACGAGAT',  # Read2的3'端适配器\n",
    "#             '-g', 'AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC',  # Read1的5'端适配器\n",
    "#             '-G', 'AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTAGATCTCGGTGGTCGCCGTATCATT',  # Read2的5'端适配器\n",
    "#             '-o', fq1_val_path,  # 修剪后的Read1输出文件\n",
    "#             '-p', fq2_val_path,  # 修剪后的Read2输出文件\n",
    "#             '--pair-filter=any',  # 如果任一读段被修剪，则保留配对读段\n",
    "#             '-e', '0.1',  # 允许的最大错误率\n",
    "#             fq1_path,  # Read1的输入文件\n",
    "#             fq2_path   # Read2的输入文件\n",
    "#         ]\n",
    "\n",
    "#         # 执行 cutadapt 命令\n",
    "#         try:\n",
    "#             process = subprocess.run(cutadapt_command, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
    "#             print(process.stdout.decode())\n",
    "#             print(process.stderr.decode())\n",
    "#         except subprocess.CalledProcessError as e:\n",
    "#             print(f\"cutadapt command failed with error: {e}\")\n",
    "#         except Exception as e:\n",
    "#             print(f\"An unexpected error occurred: {e}\")\n",
    "\n",
    "# trim_galore\n",
    "with open('file_all', 'r') as f:\n",
    "    for line in f:\n",
    "        fq1, fq2 = line.strip().split('\\t')\n",
    "        fq1_path = os.path.join(work_dir, fq1)\n",
    "        fq2_path = os.path.join(work_dir, fq2)\n",
    "\n",
    "        command = f'trim_galore -q 20 --phred33 --length 20 -e 0.1 --stringency 3 --paired --gzip -o {output_dir} {fq1_path} {fq2_path}'\n",
    "\n",
    "        try:\n",
    "            process = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
    "            stdout, stderr = process.communicate()\n",
    "            print(stdout.decode())\n",
    "            print(stderr.decode())\n",
    "        except subprocess.CalledProcessError as e:\n",
    "            print(f\"trim_galore command failed with error: {e}\")\n",
    "        except Exception as e:\n",
    "            print(f\"An unexpected error occurred: {e}\")\n",
    "\n",
    "\n",
    "## 单端\n",
    "# fq1_files = [file for file in os.listdir(work_dir) if 'EcST1.fq.gz' in file]\n",
    "# # 生成文件列表\n",
    "# file_all = '\\n'.join(fq1 for fq1 in fq1_files)\n",
    "# with open('file_all', 'w') as f:\n",
    "#     f.write(file_all)\n",
    "\n",
    "# # 处理每个 _R1.fq.gz 文件\n",
    "# with open('file_all', 'r') as f:\n",
    "#     for line in f:\n",
    "#         fq1 = line.strip()\n",
    "#         fq1_path = os.path.join(work_dir, fq1)\n",
    "\n",
    "#         # 构建 trim_galore 命令\n",
    "#         command = f'trim_galore -q 20 --phred33 --length 20 -e 0.1 --stringency 3 --gzip -o {output_dir} {fq1_path}'    ## --length 20\n",
    "\n",
    "#         try:\n",
    "#             # 执行命令\n",
    "#             process = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
    "#             stdout, stderr = process.communicate()\n",
    "#             print(stdout.decode())\n",
    "#             print(stderr.decode())\n",
    "#         except subprocess.CalledProcessError as e:\n",
    "#             print(f\"trim_galore command failed with error: {e}\")\n",
    "#         except Exception as e:\n",
    "#             print(f\"An unexpected error occurred: {e}\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 44 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 44 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 44 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 7, merged 20; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 3, merged 5; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 1; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 0; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 35596.1 (target: 50057)\n",
      "Getting block 1 of 30\n",
      "  Reserving size (50058) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "Getting block 2 of 30\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  Reserving size (50058) for bucket 2\n",
      "Getting block 3 of 30\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Reserving size (50058) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "Getting block 4 of 30\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  Reserving size (50058) for bucket 4\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 3: 10%\n",
      "  bucket 1: 10%\n",
      "  bucket 4: 10%\n",
      "  bucket 2: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 4: 20%\n",
      "  bucket 1: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 2: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 1: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 2: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 1: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 2: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 1: 50%\n",
      "  bucket 3: 70%\n",
      "  bucket 4: 60%\n",
      "  bucket 1: 60%\n",
      "  bucket 2: 50%\n",
      "  bucket 3: 80%\n",
      "  bucket 4: 70%\n",
      "  bucket 1: 70%\n",
      "  bucket 3: 90%\n",
      "  bucket 2: 60%\n",
      "  bucket 4: 80%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 30099 for bucket 3\n",
      "  (Using difference cover)\n",
      "  bucket 1: 80%\n",
      "  bucket 2: 70%\n",
      "  bucket 4: 90%\n",
      "  bucket 1: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 43819 for bucket 4\n",
      "  (Using difference cover)\n",
      "  bucket 2: 80%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 31763 for bucket 1\n",
      "  (Using difference cover)\n",
      "  bucket 2: 90%\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 30100 for bucket 3\n",
      "Getting block 5 of 30\n",
      "  Reserving size (50058) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 38630 for bucket 2\n",
      "  (Using difference cover)\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 43820 for bucket 4\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 31764 for bucket 1\n",
      "  bucket 5: 30%\n",
      "Getting block 6 of 30\n",
      "  Reserving size (50058) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 5: 40%\n",
      "Getting block 7 of 30\n",
      "  Reserving size (50058) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 6: 10%\n",
      "  bucket 5: 50%\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 38631 for bucket 2\n",
      "  bucket 7: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 5: 60%\n",
      "Getting block 8 of 30\n",
      "  Reserving size (50058) for bucket 8\n",
      "  Calculating Z arrays for bucket 8\n",
      "  Entering block accumulator loop for bucket 8:\n",
      "  bucket 5: 70%\n",
      "  bucket 6: 30%\n",
      "  bucket 7: 20%\n",
      "  bucket 8: 10%\n",
      "  bucket 5: 80%\n",
      "  bucket 6: 40%\n",
      "  bucket 5: 90%\n",
      "  bucket 6: 50%\n",
      "  bucket 7: 30%\n",
      "  bucket 8: 20%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 41135 for bucket 5\n",
      "  (Using difference cover)\n",
      "  bucket 6: 60%\n",
      "  bucket 7: 40%\n",
      "  bucket 8: 30%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 7: 50%\n",
      "  bucket 8: 40%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 38933 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 41136 for bucket 5\n",
      "  bucket 7: 60%\n",
      "  bucket 8: 50%\n",
      "Getting block 9 of 30\n",
      "  Reserving size (50058) for bucket 9\n",
      "  Calculating Z arrays for bucket 9\n",
      "  Entering block accumulator loop for bucket 9:\n",
      "  bucket 8: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 9: 10%\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 38934 for bucket 6\n",
      "  bucket 8: 70%\n",
      "  bucket 7: 80%\n",
      "Getting block 10 of 30\n",
      "  Reserving size (50058) for bucket 10\n",
      "  Calculating Z arrays for bucket 10\n",
      "  Entering block accumulator loop for bucket 10:\n",
      "  bucket 9: 20%\n",
      "  bucket 8: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 9: 30%\n",
      "  bucket 10: 10%\n",
      "  bucket 8: 90%\n",
      "  bucket 9: 40%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 39454 for bucket 7\n",
      "  (Using difference cover)\n",
      "  bucket 10: 20%\n",
      "  bucket 8: 100%\n",
      "  Sorting block of length 30857 for bucket 8\n",
      "  (Using difference cover)\n",
      "  bucket 9: 50%\n",
      "  bucket 10: 30%\n",
      "  bucket 9: 60%\n",
      "  bucket 10: 40%\n",
      "  bucket 9: 70%\n",
      "  bucket 10: 50%\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 39455 for bucket 7\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 30858 for bucket 8\n",
      "Getting block 11 of 30\n",
      "  Reserving size (50058) for bucket 11\n",
      "  Calculating Z arrays for bucket 11\n",
      "  Entering block accumulator loop for bucket 11:\n",
      "Getting block 12 of 30\n",
      "  Reserving size (50058) for bucket 12\n",
      "  Calculating Z arrays for bucket 12\n",
      "  Entering block accumulator loop for bucket 12:\n",
      "  bucket 9: 80%\n",
      "  bucket 10: 60%\n",
      "  bucket 11: 10%\n",
      "  bucket 9: 90%\n",
      "  bucket 10: 70%\n",
      "  bucket 12: 10%\n",
      "  bucket 11: 20%\n",
      "  bucket 9: 100%\n",
      "  Sorting block of length 28294 for bucket 9\n",
      "  (Using difference cover)\n",
      "  bucket 10: 80%\n",
      "  bucket 12: 20%\n",
      "  bucket 11: 30%\n",
      "  bucket 10: 90%\n",
      "  bucket 12: 30%\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 28295 for bucket 9\n",
      "  bucket 10: 100%\n",
      "  bucket 11: 40%\n",
      "  Sorting block of length 36183 for bucket 10\n",
      "  (Using difference cover)\n",
      "Getting block 13 of 30\n",
      "  bucket 12: 40%\n",
      "  Reserving size (50058) for bucket 13\n",
      "  Calculating Z arrays for bucket 13\n",
      "  Entering block accumulator loop for bucket 13:\n",
      "  bucket 11: 50%\n",
      "  bucket 12: 50%\n",
      "  bucket 13: 10%\n",
      "  bucket 11: 60%\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 36184 for bucket 10\n",
      "  bucket 12: 60%\n",
      "  bucket 11: 70%\n",
      "  bucket 13: 20%\n",
      "Getting block 14 of 30\n",
      "  Reserving size (50058) for bucket 14\n",
      "  Calculating Z arrays for bucket 14\n",
      "  Entering block accumulator loop for bucket 14:\n",
      "  bucket 12: 70%\n",
      "  bucket 11: 80%\n",
      "  bucket 14: 10%\n",
      "  bucket 13: 30%\n",
      "  bucket 12: 80%\n",
      "  bucket 11: 90%\n",
      "  bucket 14: 20%\n",
      "  bucket 13: 40%\n",
      "  bucket 12: 90%\n",
      "  bucket 11: 100%\n",
      "  Sorting block of length 41728 for bucket 11\n",
      "  (Using difference cover)\n",
      "  bucket 14: 30%\n",
      "  bucket 13: 50%\n",
      "  bucket 12: 100%\n",
      "  Sorting block of length 45263 for bucket 12\n",
      "  (Using difference cover)\n",
      "  bucket 14: 40%\n",
      "  bucket 13: 60%\n",
      "  bucket 14: 50%\n",
      "  bucket 13: 70%\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 41729 for bucket 11\n",
      "  bucket 14: 60%\n",
      "Getting block 15 of 30\n",
      "  Reserving size (50058) for bucket 15\n",
      "  Calculating Z arrays for bucket 15\n",
      "  Entering block accumulator loop for bucket 15:\n",
      "  bucket 13: 80%\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 45264 for bucket 12\n",
      "  bucket 14: 70%\n",
      "  bucket 15: 10%\n",
      "Getting block 16 of 30\n",
      "  Reserving size (50058) for bucket 16\n",
      "  Calculating Z arrays for bucket 16\n",
      "  Entering block accumulator loop for bucket 16:\n",
      "  bucket 13: 90%\n",
      "  bucket 14: 80%\n",
      "  bucket 15: 20%\n",
      "  bucket 16: 10%\n",
      "  bucket 13: 100%\n",
      "  bucket 14: 90%\n",
      "  Sorting block of length 47239 for bucket 13\n",
      "  (Using difference cover)\n",
      "  bucket 15: 30%\n",
      "  bucket 16: 20%\n",
      "  bucket 14: 100%\n",
      "  Sorting block of length 32005 for bucket 14\n",
      "  (Using difference cover)\n",
      "  bucket 15: 40%\n",
      "  bucket 16: 30%\n",
      "  bucket 15: 50%\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 47240 for bucket 13\n",
      "  bucket 16: 40%\n",
      "Getting block 17 of 30\n",
      "  Reserving size (50058) for bucket 17\n",
      "  Calculating Z arrays for bucket 17\n",
      "  Entering block accumulator loop for bucket 17:\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 32006 for bucket 14\n",
      "  bucket 15: 60%\n",
      "  bucket 16: 50%\n",
      "Getting block 18 of 30\n",
      "  Reserving size (50058) for bucket 18\n",
      "  Calculating Z arrays for bucket 18\n",
      "  Entering block accumulator loop for bucket 18:\n",
      "  bucket 17: 10%\n",
      "  bucket 15: 70%\n",
      "  bucket 16: 60%\n",
      "  bucket 18: 10%\n",
      "  bucket 17: 20%\n",
      "  bucket 15: 80%\n",
      "  bucket 16: 70%\n",
      "  bucket 18: 20%\n",
      "  bucket 17: 30%\n",
      "  bucket 15: 90%\n",
      "  bucket 16: 80%\n",
      "  bucket 18: 30%\n",
      "  bucket 17: 40%\n",
      "  bucket 15: 100%\n",
      "  Sorting block of length 39675 for bucket 15\n",
      "  (Using difference cover)\n",
      "  bucket 16: 90%\n",
      "  bucket 17: 50%\n",
      "  bucket 18: 40%\n",
      "  bucket 16: 100%\n",
      "  Sorting block of length 44985 for bucket 16\n",
      "  (Using difference cover)\n",
      "  bucket 17: 60%\n",
      "  bucket 18: 50%\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 39676 for bucket 15\n",
      "  bucket 17: 70%\n",
      "  bucket 18: 60%\n",
      "Getting block 19 of 30\n",
      "  Reserving size (50058) for bucket 19\n",
      "  Calculating Z arrays for bucket 19\n",
      "  Entering block accumulator loop for bucket 19:\n",
      "  bucket 17: 80%\n",
      "  bucket 18: 70%\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 44986 for bucket 16\n",
      "  bucket 19: 10%\n",
      "Getting block 20 of 30\n",
      "  Reserving size (50058) for bucket 20\n",
      "  Calculating Z arrays for bucket 20\n",
      "  Entering block accumulator loop for bucket 20:\n",
      "  bucket 17: 90%\n",
      "  bucket 18: 80%\n",
      "  bucket 19: 20%\n",
      "  bucket 20: 10%\n",
      "  bucket 17: 100%\n",
      "  Sorting block of length 7118 for bucket 17\n",
      "  (Using difference cover)\n",
      "  bucket 18: 90%\n",
      "  bucket 19: 30%\n",
      "  bucket 20: 20%\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 7119 for bucket 17\n",
      "Getting block 21 of 30\n",
      "  Reserving size (50058) for bucket 21\n",
      "  Calculating Z arrays for bucket 21\n",
      "  Entering block accumulator loop for bucket 21:\n",
      "  bucket 18: 100%\n",
      "  Sorting block of length 46562 for bucket 18\n",
      "  (Using difference cover)\n",
      "  bucket 19: 40%\n",
      "  bucket 20: 30%\n",
      "  bucket 21: 10%\n",
      "  bucket 19: 50%\n",
      "  bucket 20: 40%\n",
      "  bucket 21: 20%\n",
      "  bucket 19: 60%\n",
      "  bucket 20: 50%\n",
      "  bucket 19: 70%\n",
      "  bucket 21: 30%\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 46563 for bucket 18\n",
      "  bucket 20: 60%\n",
      "  bucket 19: 80%\n",
      "  bucket 21: 40%\n",
      "Getting block 22 of 30\n",
      "  Reserving size (50058) for bucket 22\n",
      "  Calculating Z arrays for bucket 22\n",
      "  Entering block accumulator loop for bucket 22:\n",
      "  bucket 20: 70%\n",
      "  bucket 19: 90%\n",
      "  bucket 21: 50%\n",
      "  bucket 22: 10%\n",
      "  bucket 20: 80%\n",
      "  bucket 19: 100%\n",
      "  Sorting block of length 23450 for bucket 19\n",
      "  (Using difference cover)\n",
      "  bucket 21: 60%\n",
      "  bucket 22: 20%\n",
      "  bucket 20: 90%\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 23451 for bucket 19\n",
      "  bucket 22: 30%\n",
      "  bucket 21: 70%\n",
      "Getting block 23 of 30\n",
      "  Reserving size (50058) for bucket 23\n",
      "  Calculating Z arrays for bucket 23\n",
      "  Entering block accumulator loop for bucket 23:\n",
      "  bucket 20: 100%\n",
      "  Sorting block of length 49369 for bucket 20\n",
      "  (Using difference cover)\n",
      "  bucket 22: 40%\n",
      "  bucket 21: 80%\n",
      "  bucket 23: 10%\n",
      "  bucket 22: 50%\n",
      "  bucket 23: 20%\n",
      "  bucket 21: 90%\n",
      "  bucket 22: 60%\n",
      "  bucket 23: 30%\n",
      "  bucket 21: 100%\n",
      "  Sorting block of length 28279 for bucket 21\n",
      "  (Using difference cover)\n",
      "  bucket 22: 70%\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 49370 for bucket 20\n",
      "  bucket 23: 40%\n",
      "  bucket 22: 80%\n",
      "Getting block 24 of 30\n",
      "  Reserving size (50058) for bucket 24\n",
      "  Calculating Z arrays for bucket 24\n",
      "  Entering block accumulator loop for bucket 24:\n",
      "  bucket 23: 50%\n",
      "  bucket 22: 90%\n",
      "  bucket 24: 10%\n",
      "  bucket 23: 60%\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 28280 for bucket 21\n",
      "  bucket 22: 100%\n",
      "  Sorting block of length 38519 for bucket 22\n",
      "  (Using difference cover)\n",
      "Getting block 25 of 30\n",
      "  Reserving size (50058) for bucket 25\n",
      "  Calculating Z arrays for bucket 25\n",
      "  Entering block accumulator loop for bucket 25:\n",
      "  bucket 24: 20%\n",
      "  bucket 23: 70%\n",
      "  bucket 25: 10%\n",
      "  bucket 24: 30%\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 38520 for bucket 22\n",
      "  bucket 23: 80%\n",
      "Getting block 26 of 30\n",
      "  Reserving size (50058) for bucket 26\n",
      "  Calculating Z arrays for bucket 26\n",
      "  Entering block accumulator loop for bucket 26:\n",
      "  bucket 24: 40%\n",
      "  bucket 25: 20%\n",
      "  bucket 23: 90%\n",
      "  bucket 26: 10%\n",
      "  bucket 24: 50%\n",
      "  bucket 25: 30%\n",
      "  bucket 23: 100%\n",
      "  Sorting block of length 37997 for bucket 23\n",
      "  (Using difference cover)\n",
      "  bucket 26: 20%\n",
      "  bucket 24: 60%\n",
      "  bucket 25: 40%\n",
      "  bucket 26: 30%\n",
      "  bucket 24: 70%\n",
      "  bucket 25: 50%\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 37998 for bucket 23\n",
      "  bucket 26: 40%\n",
      "  bucket 24: 80%\n",
      "  bucket 25: 60%\n",
      "Getting block 27 of 30\n",
      "  Reserving size (50058) for bucket 27\n",
      "  Calculating Z arrays for bucket 27\n",
      "  Entering block accumulator loop for bucket 27:\n",
      "  bucket 26: 50%\n",
      "  bucket 24: 90%\n",
      "  bucket 25: 70%\n",
      "  bucket 27: 10%\n",
      "  bucket 26: 60%\n",
      "  bucket 24: 100%\n",
      "  Sorting block of length 18712 for bucket 24\n",
      "  (Using difference cover)\n",
      "  bucket 25: 80%\n",
      "  bucket 27: 20%\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 18713 for bucket 24\n",
      "  bucket 26: 70%\n",
      "Getting block 28 of 30\n",
      "  Reserving size (50058) for bucket 28\n",
      "  Calculating Z arrays for bucket 28\n",
      "  Entering block accumulator loop for bucket 28:\n",
      "  bucket 25: 90%\n",
      "  bucket 27: 30%\n",
      "  bucket 26: 80%\n",
      "  bucket 28: 10%\n",
      "  bucket 25: 100%\n",
      "  Sorting block of length 40469 for bucket 25\n",
      "  (Using difference cover)\n",
      "  bucket 27: 40%\n",
      "  bucket 26: 90%\n",
      "  bucket 28: 20%\n",
      "  bucket 27: 50%\n",
      "  bucket 26: 100%\n",
      "  Sorting block of length 20032 for bucket 26\n",
      "  (Using difference cover)\n",
      "  bucket 28: 30%\n",
      "  bucket 27: 60%\n",
      "  bucket 28: 40%\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 20033 for bucket 26\n",
      "  bucket 27: 70%\n",
      "  bucket 28: 50%\n",
      "Getting block 29 of 30\n",
      "  Reserving size (50058) for bucket 29\n",
      "  Calculating Z arrays for bucket 29\n",
      "  Entering block accumulator loop for bucket 29:\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 40470 for bucket 25\n",
      "  bucket 27: 80%\n",
      "  bucket 28: 60%\n",
      "Getting block 30 of 30\n",
      "  Reserving size (50058) for bucket 30\n",
      "  Calculating Z arrays for bucket 30\n",
      "  Entering block accumulator loop for bucket 30:\n",
      "  bucket 29: 10%\n",
      "  bucket 27: 90%\n",
      "  bucket 30: 10%\n",
      "  bucket 29: 20%\n",
      "  bucket 28: 70%\n",
      "  bucket 30: 20%\n",
      "  bucket 27: 100%\n",
      "  Sorting block of length 49370 for bucket 27\n",
      "  (Using difference cover)\n",
      "  bucket 29: 30%\n",
      "  bucket 30: 30%\n",
      "  bucket 28: 80%\n",
      "  bucket 29: 40%\n",
      "  bucket 30: 40%\n",
      "  bucket 28: 90%\n",
      "  bucket 30: 50%\n",
      "  bucket 29: 50%\n",
      "  bucket 30: 60%\n",
      "  bucket 28: 100%\n",
      "  Sorting block of length 38201 for bucket 28\n",
      "  (Using difference cover)\n",
      "  bucket 29: 60%\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 49371 for bucket 27\n",
      "  bucket 30: 70%\n",
      "  bucket 29: 70%\n",
      "  bucket 30: 80%\n",
      "  bucket 30: 90%\n",
      "  bucket 29: 80%\n",
      "  bucket 30: 100%\n",
      "  Sorting block of length 40217 for bucket 30\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 38202 for bucket 28\n",
      "  bucket 29: 90%\n",
      "  bucket 29: 100%\n",
      "  Sorting block of length 19525 for bucket 29\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 40218 for bucket 30\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 19526 for bucket 29\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## 构建索引\n",
    "import os\n",
    "import subprocess\n",
    "\n",
    "# hisat2\n",
    "# geneRef_dir = '/home/john/RNA-seq/RNASeqDemoData/reference/GRCh38/'\n",
    "# geneRef_name = 'GRCh38.genome'\n",
    "# geneRef_fa = geneRef_dir + geneRef_name + '.fa'\n",
    "# command = ['hisat2-build', '-p', '4', geneRef_fa, geneRef_name]\n",
    "\n",
    "# try:\n",
    "#     process = subprocess.run(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, check=True)\n",
    "#     print(process.stdout.decode())\n",
    "# except subprocess.CalledProcessError as e:\n",
    "#     print(f\"bwa index command failed with error: {e}\")\n",
    "\n",
    "\n",
    "# ## bwa\n",
    "\n",
    "# ref_genome_path = '/home/john/ZouLab_Code/RNASeq/reference/miRNA/processed_sequences.fa'\n",
    "# output_dir = '/home/john/ZouLab_Code/RNASeq/reference/GCF5845_bacter_bwa_index'\n",
    "# os.makedirs(output_dir, exist_ok=True)\n",
    "# command = ['bwa', 'index', ref_genome_path]\n",
    "\n",
    "# try:\n",
    "#     process = subprocess.run(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, check=True)\n",
    "#     print(process.stdout.decode())\n",
    "# except subprocess.CalledProcessError as e:\n",
    "#     print(f\"bwa index command failed with error: {e}\")\n",
    "\n",
    "\n",
    "## star\n",
    "# thread_count = 5\n",
    "# star_command = [\n",
    "#     'STAR',\n",
    "#     '--runMode', 'genomeGenerate',\n",
    "#     '--runThreadN', str(thread_count),\n",
    "#     '--genomeDir', '/home/john/RNA-seq/RNASeqDemoData/reference/star_index',\n",
    "#     '--genomeFastaFiles', '/home/john/RNA-seq/RNASeqDemoData/reference/GRCh38/GRCh38.genome.fa',\n",
    "#     '--sjdbGTFfile', '/home/john/RNA-seq/RNASeqDemoData/reference/GRCh38/gencode.v41.annotation.gtf',\n",
    "#     '--sjdbOverhang', '149'\n",
    "# ]\n",
    "\n",
    "# process = subprocess.run(star_command, stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
    "# print(process.stdout.decode())\n",
    "# print(process.stderr.decode())\n",
    "\n",
    "\n",
    "## hisat2\n",
    "geneRef_dir = '/home/john/ZouLab_Code/RNASeq/reference/miRNA/'\n",
    "geneRef_name = 'processed_sequences'\n",
    "geneRef_fa = geneRef_dir + geneRef_name + '.fa'\n",
    "command = ['hisat2-build', '-p', '4', geneRef_fa, geneRef_name]\n",
    "\n",
    "try:\n",
    "    process = subprocess.run(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, check=True)\n",
    "    print(process.stdout.decode())\n",
    "except subprocess.CalledProcessError as e:\n",
    "    print(f\"bwa index command failed with error: {e}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# alignment (hisat2 / star / bwa)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hisat2 success\n",
      "CompletedProcess(args=['hisat2', '-p', '10', '-x', '/home/john/ZouLab_Code/RNASeq/reference/hisat2index/GRCh38.p14.genome', '-1', '/home/john/ZouLab_Code/RNASeq/TrimGaloreData/HPAdata/HPAA1_R1_val_1.fq.gz', '-2', '/home/john/ZouLab_Code/RNASeq/TrimGaloreData/HPAdata/HPAA1_R1_val_1.fq.gz', '-S', '/home/john/ZouLab_Code/RNASeq/samData/sam_HPA_GRCh38p14/HPAA1.sam'], returncode=0, stdout='')\n",
      "Hisat2 success\n",
      "CompletedProcess(args=['hisat2', '-p', '10', '-x', '/home/john/ZouLab_Code/RNASeq/reference/hisat2index/GRCh38.p14.genome', '-1', '/home/john/ZouLab_Code/RNASeq/TrimGaloreData/HPAdata/HPAA2_R1_val_1.fq.gz', '-2', '/home/john/ZouLab_Code/RNASeq/TrimGaloreData/HPAdata/HPAA2_R1_val_1.fq.gz', '-S', '/home/john/ZouLab_Code/RNASeq/samData/sam_HPA_GRCh38p14/HPAA2.sam'], returncode=0, stdout='')\n",
      "Hisat2 success\n",
      "CompletedProcess(args=['hisat2', '-p', '10', '-x', '/home/john/ZouLab_Code/RNASeq/reference/hisat2index/GRCh38.p14.genome', '-1', '/home/john/ZouLab_Code/RNASeq/TrimGaloreData/HPAdata/HPAA3_R1_val_1.fq.gz', '-2', '/home/john/ZouLab_Code/RNASeq/TrimGaloreData/HPAdata/HPAA3_R1_val_1.fq.gz', '-S', '/home/john/ZouLab_Code/RNASeq/samData/sam_HPA_GRCh38p14/HPAA3.sam'], returncode=0, stdout='')\n",
      "Hisat2 success\n",
      "CompletedProcess(args=['hisat2', '-p', '10', '-x', '/home/john/ZouLab_Code/RNASeq/reference/hisat2index/GRCh38.p14.genome', '-1', '/home/john/ZouLab_Code/RNASeq/TrimGaloreData/HPAdata/HPAD1_R1_val_1.fq.gz', '-2', '/home/john/ZouLab_Code/RNASeq/TrimGaloreData/HPAdata/HPAD1_R1_val_1.fq.gz', '-S', '/home/john/ZouLab_Code/RNASeq/samData/sam_HPA_GRCh38p14/HPAD1.sam'], returncode=0, stdout='')\n",
      "Hisat2 success\n",
      "CompletedProcess(args=['hisat2', '-p', '10', '-x', '/home/john/ZouLab_Code/RNASeq/reference/hisat2index/GRCh38.p14.genome', '-1', '/home/john/ZouLab_Code/RNASeq/TrimGaloreData/HPAdata/HPAD2_R1_val_1.fq.gz', '-2', '/home/john/ZouLab_Code/RNASeq/TrimGaloreData/HPAdata/HPAD2_R1_val_1.fq.gz', '-S', '/home/john/ZouLab_Code/RNASeq/samData/sam_HPA_GRCh38p14/HPAD2.sam'], returncode=0, stdout='')\n",
      "Hisat2 success\n",
      "CompletedProcess(args=['hisat2', '-p', '10', '-x', '/home/john/ZouLab_Code/RNASeq/reference/hisat2index/GRCh38.p14.genome', '-1', '/home/john/ZouLab_Code/RNASeq/TrimGaloreData/HPAdata/HPAD3_R1_val_1.fq.gz', '-2', '/home/john/ZouLab_Code/RNASeq/TrimGaloreData/HPAdata/HPAD3_R1_val_1.fq.gz', '-S', '/home/john/ZouLab_Code/RNASeq/samData/sam_HPA_GRCh38p14/HPAD3.sam'], returncode=0, stdout='')\n",
      "Hisat2 success\n",
      "CompletedProcess(args=['hisat2', '-p', '10', '-x', '/home/john/ZouLab_Code/RNASeq/reference/hisat2index/GRCh38.p14.genome', '-1', '/home/john/ZouLab_Code/RNASeq/TrimGaloreData/HPAdata/HPAT1_R1_val_1.fq.gz', '-2', '/home/john/ZouLab_Code/RNASeq/TrimGaloreData/HPAdata/HPAT1_R1_val_1.fq.gz', '-S', '/home/john/ZouLab_Code/RNASeq/samData/sam_HPA_GRCh38p14/HPAT1.sam'], returncode=0, stdout='')\n",
      "Hisat2 success\n",
      "CompletedProcess(args=['hisat2', '-p', '10', '-x', '/home/john/ZouLab_Code/RNASeq/reference/hisat2index/GRCh38.p14.genome', '-1', '/home/john/ZouLab_Code/RNASeq/TrimGaloreData/HPAdata/HPAT2_R1_val_1.fq.gz', '-2', '/home/john/ZouLab_Code/RNASeq/TrimGaloreData/HPAdata/HPAT2_R1_val_1.fq.gz', '-S', '/home/john/ZouLab_Code/RNASeq/samData/sam_HPA_GRCh38p14/HPAT2.sam'], returncode=0, stdout='')\n",
      "Hisat2 success\n",
      "CompletedProcess(args=['hisat2', '-p', '10', '-x', '/home/john/ZouLab_Code/RNASeq/reference/hisat2index/GRCh38.p14.genome', '-1', '/home/john/ZouLab_Code/RNASeq/TrimGaloreData/HPAdata/HPAT3_R1_val_1.fq.gz', '-2', '/home/john/ZouLab_Code/RNASeq/TrimGaloreData/HPAdata/HPAT3_R1_val_1.fq.gz', '-S', '/home/john/ZouLab_Code/RNASeq/samData/sam_HPA_GRCh38p14/HPAT3.sam'], returncode=0, stdout='')\n"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "\n",
    "cleaned_fq_dir = '/home/john/ZouLab_Code/RNASeq/TrimGaloreData/HPAdata/'\n",
    "outputSam_dir = '/home/john/ZouLab_Code/RNASeq/samData/sam_HPA_GRCh38p14/'\n",
    "outputLog_dir = '/home/john/ZouLab_Code/RNASeq/samData/sam_HPA_GRCh38p14/SamLog/'\n",
    "# sampleAll = ['HeSRA1', 'HeSRA2', 'HeSRA3', 'HeSRD1', 'HeSRD2', 'HeSRD3', 'HeSRT1', 'HeSRT2', 'HeSRT3']\n",
    "# 'H4RIP-2', 'H4RIP-3', 'H4RIP-5', 'H4RIP-6', 'H4RIP-7'\n",
    "sampleAll = ['HPAA1', 'HPAA2', 'HPAA3', 'HPAD1', 'HPAD2', 'HPAD3', 'HPAT1', 'HPAT2', 'HPAT3']\n",
    "## hisat2 双端\n",
    "for sample in sampleAll:\n",
    "    command = [\n",
    "        'hisat2',\n",
    "        '-p', '10',\n",
    "        '-x', '/home/john/ZouLab_Code/RNASeq/reference/hisat2index/GRCh38.p14.genome', #'/home/john/RNA-seq/RNASeqDemoData/reference/ht_index/GRCh38.genome'\n",
    "        '-1', cleaned_fq_dir + sample + '_R1_val_1.fq.gz',\n",
    "        '-2', cleaned_fq_dir + sample + '_R1_val_1.fq.gz',\n",
    "        '-S', outputSam_dir + sample + '.sam'\n",
    "    ]\n",
    "\n",
    "    output_log = outputLog_dir + sample + '.log'\n",
    "\n",
    "    with open(output_log, 'w') as log:\n",
    "        try:\n",
    "            result = subprocess.run(command, stdout=subprocess.PIPE, stderr=log, check=True, text=True)\n",
    "            print(\"Hisat2 success\")\n",
    "            print(result)\n",
    "        except subprocess.CalledProcessError as e:\n",
    "            print(f\"Hisat2 failed with error: {e}\")\n",
    "\n",
    "\n",
    "# hisat2单端\n",
    "# for sample in sampleAll:\n",
    "#     command = [\n",
    "#         'hisat2',\n",
    "#         '-p', '5',  # 指定使用10个线程\n",
    "#         '-x', '/home/john/ZouLab_Code/RNASeq/reference/GCF5845_bacter_index/GCF_000005845.2_ASM584v2_genomic', #'/home/john/RNA-seq/RNASeqDemoData/reference/ht_index/GRCh38.genome'\n",
    "#         '-U', cleaned_fq_dir + sample + '_trimmed.fq.gz',\n",
    "#         '-S', outputSam_dir + sample + '.sam'\n",
    "#     ]\n",
    "\n",
    "#     output_log = outputLog_dir + sample + '.log'\n",
    "\n",
    "#     with open(output_log, 'w') as log:\n",
    "#         try:\n",
    "#             result = subprocess.run(command, stdout=subprocess.PIPE, stderr=log, check=True, text=True)\n",
    "#             print(\"Hisat2 success\")\n",
    "#             print(result)\n",
    "#         except subprocess.CalledProcessError as e:\n",
    "#             print(f\"Hisat2 failed with error: {e}\")\n",
    "\n",
    "\n",
    "## bwa\n",
    "\n",
    "# cleaned_fq_dir = '/home/john/ZouLab_Code/RNASeq/TrimGaloreData/HESR_TX114_cleanedData_Bw/'\n",
    "# outputSam_dir = '/home/john/ZouLab_Code/RNASeq/samData/sam_bwa_HESR_TX114_cleanedData_Bw_GCF/'\n",
    "# outputLog_dir = '/home/john/ZouLab_Code/RNASeq/samData/sam_bwa_HESR_TX114_cleanedData_Bw_GCF/SamLog/'\n",
    "# sampleAll = ['BwA3', 'BwD2', 'BwT1']\n",
    "# for sample in sampleAll:\n",
    "#     command = [\n",
    "#         'bwa', 'mem',\n",
    "#         '-t', '5',  ## 5线程\n",
    "#         '-M', \n",
    "#         '/home/john/ZouLab_Code/RNASeq/reference/GCF5845_bacter_bwa_index/GCF_000005845.2_ASM584v2_genomic.fna',\n",
    "#         cleaned_fq_dir + sample + '_R1_val_1.fq.gz',\n",
    "#         cleaned_fq_dir + sample + '_R2_val_2.fq.gz',\n",
    "#         ]\n",
    "#     with open(outputSam_dir + sample + '.bwa.sam', 'w') as output_file:\n",
    "#         try:\n",
    "#             subprocess.run(command, stdout=output_file, stderr=subprocess.PIPE, check=True)\n",
    "#             print(\"bwa success\")\n",
    "#         except subprocess.CalledProcessError as e:\n",
    "#             print(f\"bwa failed with error: {e}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# transf sam to bam (samtools for bwa and hisat2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CompletedProcess(args=['samtools', 'sort', '-O', 'bam', '-@', '10', '-o', '/home/john/ZouLab_Code/RNASeq/bamData/bam_HPA_GRCh38p14/HPAA2.bam', '/home/john/ZouLab_Code/RNASeq/samData/sam_HPA_GRCh38p14/HPAA2.sam'], returncode=0, stdout=b'', stderr=b'[bam_sort_core] merging from 40 files and 10 in-memory blocks...\\n')\n",
      "CompletedProcess(args=['samtools', 'index', '/home/john/ZouLab_Code/RNASeq/bamData/bam_HPA_GRCh38p14/HPAA2.bam'], returncode=0, stdout=b'', stderr=b'')\n",
      "CompletedProcess(args=['samtools', 'flagstat', '-@', '10', '/home/john/ZouLab_Code/RNASeq/bamData/bam_HPA_GRCh38p14/HPAA2.bam'], returncode=0, stderr=b'')\n",
      "CompletedProcess(args=['samtools', 'sort', '-O', 'bam', '-@', '10', '-o', '/home/john/ZouLab_Code/RNASeq/bamData/bam_HPA_GRCh38p14/HPAT2.bam', '/home/john/ZouLab_Code/RNASeq/samData/sam_HPA_GRCh38p14/HPAT2.sam'], returncode=0, stdout=b'', stderr=b'[bam_sort_core] merging from 40 files and 10 in-memory blocks...\\n')\n",
      "CompletedProcess(args=['samtools', 'index', '/home/john/ZouLab_Code/RNASeq/bamData/bam_HPA_GRCh38p14/HPAT2.bam'], returncode=0, stdout=b'', stderr=b'')\n",
      "CompletedProcess(args=['samtools', 'flagstat', '-@', '10', '/home/john/ZouLab_Code/RNASeq/bamData/bam_HPA_GRCh38p14/HPAT2.bam'], returncode=0, stderr=b'')\n",
      "CompletedProcess(args=['samtools', 'sort', '-O', 'bam', '-@', '10', '-o', '/home/john/ZouLab_Code/RNASeq/bamData/bam_HPA_GRCh38p14/HPAD3.bam', '/home/john/ZouLab_Code/RNASeq/samData/sam_HPA_GRCh38p14/HPAD3.sam'], returncode=0, stdout=b'', stderr=b'[bam_sort_core] merging from 20 files and 10 in-memory blocks...\\n')\n",
      "CompletedProcess(args=['samtools', 'index', '/home/john/ZouLab_Code/RNASeq/bamData/bam_HPA_GRCh38p14/HPAD3.bam'], returncode=0, stdout=b'', stderr=b'')\n",
      "CompletedProcess(args=['samtools', 'flagstat', '-@', '10', '/home/john/ZouLab_Code/RNASeq/bamData/bam_HPA_GRCh38p14/HPAD3.bam'], returncode=0, stderr=b'')\n",
      "CompletedProcess(args=['samtools', 'sort', '-O', 'bam', '-@', '10', '-o', '/home/john/ZouLab_Code/RNASeq/bamData/bam_HPA_GRCh38p14/HPAD1.bam', '/home/john/ZouLab_Code/RNASeq/samData/sam_HPA_GRCh38p14/HPAD1.sam'], returncode=0, stdout=b'', stderr=b'[bam_sort_core] merging from 10 files and 10 in-memory blocks...\\n')\n",
      "CompletedProcess(args=['samtools', 'index', '/home/john/ZouLab_Code/RNASeq/bamData/bam_HPA_GRCh38p14/HPAD1.bam'], returncode=0, stdout=b'', stderr=b'')\n",
      "CompletedProcess(args=['samtools', 'flagstat', '-@', '10', '/home/john/ZouLab_Code/RNASeq/bamData/bam_HPA_GRCh38p14/HPAD1.bam'], returncode=0, stderr=b'')\n",
      "CompletedProcess(args=['samtools', 'sort', '-O', 'bam', '-@', '10', '-o', '/home/john/ZouLab_Code/RNASeq/bamData/bam_HPA_GRCh38p14/HPAA3.bam', '/home/john/ZouLab_Code/RNASeq/samData/sam_HPA_GRCh38p14/HPAA3.sam'], returncode=0, stdout=b'', stderr=b'[bam_sort_core] merging from 50 files and 10 in-memory blocks...\\n')\n",
      "CompletedProcess(args=['samtools', 'index', '/home/john/ZouLab_Code/RNASeq/bamData/bam_HPA_GRCh38p14/HPAA3.bam'], returncode=0, stdout=b'', stderr=b'')\n",
      "CompletedProcess(args=['samtools', 'flagstat', '-@', '10', '/home/john/ZouLab_Code/RNASeq/bamData/bam_HPA_GRCh38p14/HPAA3.bam'], returncode=0, stderr=b'')\n",
      "CompletedProcess(args=['samtools', 'sort', '-O', 'bam', '-@', '10', '-o', '/home/john/ZouLab_Code/RNASeq/bamData/bam_HPA_GRCh38p14/HPAD2.bam', '/home/john/ZouLab_Code/RNASeq/samData/sam_HPA_GRCh38p14/HPAD2.sam'], returncode=0, stdout=b'', stderr=b'[bam_sort_core] merging from 30 files and 10 in-memory blocks...\\n')\n",
      "CompletedProcess(args=['samtools', 'index', '/home/john/ZouLab_Code/RNASeq/bamData/bam_HPA_GRCh38p14/HPAD2.bam'], returncode=0, stdout=b'', stderr=b'')\n",
      "CompletedProcess(args=['samtools', 'flagstat', '-@', '10', '/home/john/ZouLab_Code/RNASeq/bamData/bam_HPA_GRCh38p14/HPAD2.bam'], returncode=0, stderr=b'')\n",
      "CompletedProcess(args=['samtools', 'sort', '-O', 'bam', '-@', '10', '-o', '/home/john/ZouLab_Code/RNASeq/bamData/bam_HPA_GRCh38p14/HPAA1.bam', '/home/john/ZouLab_Code/RNASeq/samData/sam_HPA_GRCh38p14/HPAA1.sam'], returncode=0, stdout=b'', stderr=b'[bam_sort_core] merging from 50 files and 10 in-memory blocks...\\n')\n",
      "CompletedProcess(args=['samtools', 'index', '/home/john/ZouLab_Code/RNASeq/bamData/bam_HPA_GRCh38p14/HPAA1.bam'], returncode=0, stdout=b'', stderr=b'')\n",
      "CompletedProcess(args=['samtools', 'flagstat', '-@', '10', '/home/john/ZouLab_Code/RNASeq/bamData/bam_HPA_GRCh38p14/HPAA1.bam'], returncode=0, stderr=b'')\n",
      "CompletedProcess(args=['samtools', 'sort', '-O', 'bam', '-@', '10', '-o', '/home/john/ZouLab_Code/RNASeq/bamData/bam_HPA_GRCh38p14/HPAT3.bam', '/home/john/ZouLab_Code/RNASeq/samData/sam_HPA_GRCh38p14/HPAT3.sam'], returncode=0, stdout=b'', stderr=b'[bam_sort_core] merging from 40 files and 10 in-memory blocks...\\n')\n",
      "CompletedProcess(args=['samtools', 'index', '/home/john/ZouLab_Code/RNASeq/bamData/bam_HPA_GRCh38p14/HPAT3.bam'], returncode=0, stdout=b'', stderr=b'')\n",
      "CompletedProcess(args=['samtools', 'flagstat', '-@', '10', '/home/john/ZouLab_Code/RNASeq/bamData/bam_HPA_GRCh38p14/HPAT3.bam'], returncode=0, stderr=b'')\n",
      "CompletedProcess(args=['samtools', 'sort', '-O', 'bam', '-@', '10', '-o', '/home/john/ZouLab_Code/RNASeq/bamData/bam_HPA_GRCh38p14/HPAT1.bam', '/home/john/ZouLab_Code/RNASeq/samData/sam_HPA_GRCh38p14/HPAT1.sam'], returncode=0, stdout=b'', stderr=b'[bam_sort_core] merging from 50 files and 10 in-memory blocks...\\n')\n",
      "CompletedProcess(args=['samtools', 'index', '/home/john/ZouLab_Code/RNASeq/bamData/bam_HPA_GRCh38p14/HPAT1.bam'], returncode=0, stdout=b'', stderr=b'')\n",
      "CompletedProcess(args=['samtools', 'flagstat', '-@', '10', '/home/john/ZouLab_Code/RNASeq/bamData/bam_HPA_GRCh38p14/HPAT1.bam'], returncode=0, stderr=b'')\n"
     ]
    }
   ],
   "source": [
    "## sam -> bam\n",
    "import subprocess\n",
    "import os\n",
    "\n",
    "def run_command(command, stdout=subprocess.PIPE):\n",
    "    try:\n",
    "        result = subprocess.run(command, check=True, stdout=stdout, stderr=subprocess.PIPE)\n",
    "        print(result)\n",
    "        return True\n",
    "    except subprocess.CalledProcessError as e:\n",
    "        print(f\"Command '{' '.join(command)}' failed with error: {e.stderr.decode()}\")\n",
    "        return False\n",
    "\n",
    "def sam_to_bam(input_sam, output_bam):\n",
    "    command = [\n",
    "        'samtools', 'sort',\n",
    "        '-O', 'bam',\n",
    "        '-@', '10',\n",
    "        '-o', output_bam,\n",
    "        input_sam\n",
    "    ]\n",
    "    return run_command(command)\n",
    "\n",
    "def index_bam(bam_file):\n",
    "    command = ['samtools', 'index', bam_file]\n",
    "    return run_command(command)\n",
    "\n",
    "def flagstat_bam(bam_file):\n",
    "    command = ['samtools', 'flagstat', '-@', '10', bam_file]\n",
    "    flagstat_file = bam_file.replace('.bam', '.flagstat')\n",
    "    with open(flagstat_file, 'w') as output_file:\n",
    "        return run_command(command, stdout=output_file)\n",
    "\n",
    "input_dir = '/home/john/ZouLab_Code/RNASeq/samData/sam_HPA_GRCh38p14/'\n",
    "output_dir = '/home/john/ZouLab_Code/RNASeq/bamData/bam_HPA_GRCh38p14/'\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "for file_name in os.listdir(input_dir):\n",
    "    if file_name.endswith('.sam'):\n",
    "        input_sam = os.path.join(input_dir, file_name)\n",
    "        output_bam = os.path.join(output_dir, file_name.replace('.sam', '.bam'))\n",
    "        # sam_to_bam(input_sam, output_bam)\n",
    "\n",
    "        if sam_to_bam(input_sam, output_bam):\n",
    "            index_bam(output_bam)\n",
    "            flagstat_bam(output_bam)\n",
    "\n",
    "## 这一步结束后 需要到flagstat所在的路径下，通过  multiqc ./  进行质控可视化\n",
    "## 可以通过 samtools view -h SRR1039508.bam | less 查看sort后的数据"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 基因表达定量分析 (linux 生成表达矩阵)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 转录本组装预测\n",
    "# stringtie -p 3 -e -G /home/cyh/Desktop/hugene_dir/GCF_000001405.40_GRCh38.p14_genomic.gff -o ly1.gtf -i /home/cyh/Desktop/his_result_sample1/sample1_sorted.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "63086\n"
     ]
    }
   ],
   "source": [
    "## 基因表达定量分析 (linux 生成表达矩阵) \n",
    "## 双端加 -p\n",
    "# featureCounts -T 5 -p -t exon -g gene_id -a $\"/home/john/RNA-seq/RNASeqDemoData/reference/GRCh38/gencode.v41.annotation.gtf\" -o Countstable.txt *.bam 1>counts.id.log 2>&1 &\n",
    "# GRCh38双端 : featureCounts -T 5 -p -t exon -g gene_id -a $\"/home/john/ZouLab_Code/RNASeq/reference/GRCh38.p14/gencode.v46.annotation.gtf\" -o Countstable.txt *.bam 1>counts.id.log 2>&1 &\n",
    "# GRCh38单端 : featureCounts -T 5 -t exon -g gene_id -a $\"/home/john/ZouLab_Code/RNASeq/reference/GRCh38.p14/gencode.v46.annotation.gtf\" -o Countstable.txt *.bam 1>counts.id.log 2>&1 &\n",
    "# tRNA : featureCounts -T 5 -p -t exon -g gene_id -a $\"/home/john/ZouLab_Code/RNASeq/reference/tRNA/tRNA_annotation.gtf\" -o Countstable.txt *.bam 1>counts.id.log 2>&1 &\n",
    "# miRNA : featureCounts -T 5 -t miRNA -g ID -F GFF3 -a $\"/home/john/ZouLab_Code/RNASeq/reference/miRNA/miRNA_hsa.gff3\" -o Countstable.txt *.bam 1>counts.id.log 2>&1 &\n",
    "#featureCounts -T 5 -p -t exon -g gene_id -a $\"/home/john/ZouLab_Code/RNASeq/reference/GCF5845_bacter/genomic.gtf\" -o Countstable.txt *.bam 1>counts.id.log 2>&1 &\n",
    "# cd /home/john/RNA-seq/RNASeqDemoData/featureCount\n",
    "# multiqc all.id.txt.summary\n",
    "# multiqc Countstable.txt.summary\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "prefix_dir = \"/home/john/ZouLab_Code/RNASeq/featureCountsData/FC_4HNE-RIPseq_cleanedData/\"\n",
    "countstable = \"/home/john/ZouLab_Code/RNASeq/featureCountsData/FC_4HNE-RIPseq_cleanedData/Countstable.txt\"\n",
    "\n",
    "df = pd.read_csv(countstable, sep=\"\\t\", header=0)\n",
    "df.set_index(df.columns[0], inplace=True)\n",
    "df = df.iloc[:, 0:]\n",
    "df.to_csv(prefix_dir + \"Countstable.csv\")\n",
    "\n",
    "## 只取featureCounts\n",
    "prefix_name = \"QuantAnalGeneExpression\"\n",
    "prefix = prefix_dir + prefix_name\n",
    "\n",
    "featureCounts_meta = df.iloc[:, 5:]\n",
    "featureCounts_df = featureCounts_meta[featureCounts_meta.sum(axis=1) > 0]\n",
    "featureCounts_df.to_csv(f\"{prefix_dir}FC_filted.csv\")\n",
    "\n",
    "\n",
    "# TPM计算\n",
    "print(featureCounts_meta.shape[0])\n",
    "kb = featureCounts_meta.shape[0] / 1000\n",
    "rpk = featureCounts_df / kb\n",
    "tpm = (rpk / rpk.sum()) * 1e6\n",
    "tpm.to_csv(f\"{prefix}_tpm.csv\", index_label=\"Geneid\")\n",
    "\n",
    "# # FPKM计算\n",
    "fpkm = (rpk / featureCounts_df.sum()) * 1e6\n",
    "fpkm.to_csv(f\"{prefix}_fpkm.csv\", index_label=\"Geneid\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 基因表达定量分析 (linux 生成表达矩阵)\n",
    "# featureCounts -T 5 -p -t exon -g gene_id -a $\"/home/john/RNA-seq/RNASeqDemoData/reference/GRCh38/gencode.v41.annotation.gtf\" -o Countstable.txt *.bam 1>counts.id.log 2>&1 &\n",
    "# cd /home/john/RNA-seq/RNASeqDemoData/featureCount\n",
    "# multiqc all.id.txt.summary\n",
    "\n",
    "import pandas as pd\n",
    "countstable = \"/home/john/ZouLab_Code/RNASeq/featureCountsData/FC_4HNE-RIPseq_cleanedData/Countstable.txt\"\n",
    "csv_output = \"/home/john/ZouLab_Code/RNASeq/featureCountsData/FC_4HNE-RIPseq_cleanedDataCountstable.csv\"\n",
    "df = pd.read_csv(countstable, sep=\"\\t\", header=0)\n",
    "df.to_csv(csv_output, index=False)\n",
    "\n",
    "## 只取featureCounts\n",
    "prefix_dir = \"/home/john/ZouLab_Code/RNASeq/featureCountsData/FC_4HNE-RIPseq_cleanedData/\"\n",
    "prefix_name = \"QuantAnalGeneExpression\"\n",
    "prefix = prefix_dir + prefix_name\n",
    "\n",
    "Geneid = df.iloc[:, 0:1]\n",
    "DEGNum = df.iloc[:, 6:]\n",
    "featureCounts_meta = pd.concat([Geneid, DEGNum], axis=1)\n",
    "featureCounts_df = featureCounts_meta[featureCounts_meta.iloc[:, 1:].sum(axis=1) > 0]\n",
    "featureCounts_df.to_csv(f\"{prefix_dir}FC_filted.csv\")\n",
    "\n",
    "## TPM计算\n",
    "kb = featureCounts_meta.shape[0] / 1000\n",
    "rpk = featureCounts_df / kb\n",
    "tpm = (rpk / rpk.sum()) * 1e6\n",
    "tpm.to_csv(f\"{prefix}_tpm.csv\", index_label=\"Geneid\")\n",
    "\n",
    "## FPKM计算\n",
    "fpkm = (rpk / featureCounts_df.sum()) * 1e6\n",
    "fpkm.to_csv(f\"{prefix}_fpkm.csv\", index_label=\"Geneid\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "RNASeq",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
