# Workflow of RNA-seq data analysis
This repository provides a comprehensive workflow for RNA-seq data analysis, covering everything from upstream data processing to downstream analysis. The workflow includes quality control, read alignment, gene and transcript-level quantification, visualization, differential gene expression analysis, and more. The implementation is based on methodologies described in the following literature:
- [A survey of best practices for RNA-seq data analysis](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0881-8)

The workflow is designed to be flexible and customizable, allowing users to choose from multiple algorithms for quality control and read alignment based on their project requirements.

## Contributing
Contributions are welcome! If you have suggestions for improving the workflow or adding new features, please open an issue or submit a pull request.

## Citation
If you use this workflow in your research, please cite the original paper that inspired this implementation:
Conesa, A., Madrigal, P., Tarazona, S. et al. A survey of best practices for RNA-seq data analysis. Genome Biol 17, 13 (2016). https://doi.org/10.1186/s13059-016-0881-8
